grease/src/Grease/Shape.hs

```diff
@@ -314,3 +340,270 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
 instance (MC.PrettyF tag, PrettyExt ext tag) => PP.Pretty (ArgShapes ext tag tys) where
   pretty (ArgShapes regs) =
     MC.foldlFC (\doc rShape -> PP.vcat [doc, PP.pretty rShape]) "" regs
+
+-- * Replacing
+
+data TypeMismatch
+  = TypeMismatch
+  { typeMismatchName :: String
+  , expectedType :: C.Some CT.TypeRepr
+  , foundType :: C.Some CT.TypeRepr
+  }
+  deriving Show
+
+instance PP.Pretty TypeMismatch where
+  pretty tm =
+    PP.hsep
+      [ "Type mismatch for"
+      , PP.pretty (typeMismatchName tm) PP.<> ":"
+      , "expected:"
+      , PP.viaShow (expectedType tm)
+      , "but found:"
+      , PP.viaShow (foundType tm)
+      ]
+
+-- | A mapping from argument name to the shape for that argument. Intended to be used with
+-- 'replaceShapes' to initialize members of an initial 'ArgShape' with user or elsewhere defined shapes.
+newtype ParsedShapes ext
+  = ParsedShapes {_getParsedShapes :: Map.Map Text (C.Some (Shape ext NoTag))}
+
+-- | Given an initial, provisional list of arguments and a set of replacements
+-- for some of them, calculate a new list of arguments.
+replaceShapes ::
+  forall ext w tys.
+  ExtShape ext ~ PtrShape ext w =>
+  HasPtrWidth w =>
+  -- | Argument names
+  Ctx.Assignment (Const.Const String) tys ->
+  -- | Initial arguments
+  ArgShapes ext NoTag tys ->
+  -- | Replacement arguments
+  ParsedShapes ext ->
+  Either TypeMismatch (ArgShapes ext NoTag tys)
+replaceShapes names (ArgShapes args) (ParsedShapes replacements) =
+  -- TODO: Check that all the map keys are expected
+  ArgShapes
+    <$> Ctx.zipWithM (\(Const.Const nm) s -> replaceOne nm s) names args
+ where
+  replaceOne :: String -> Shape ext NoTag t -> Either TypeMismatch (Shape ext NoTag t)
+  replaceOne nm s =
+    case Map.lookup (Text.pack nm) replacements of
+      Just (C.Some replace) ->
+        let ty = shapeType ptrShapeType s
+         in let ty' = shapeType ptrShapeType replace
+             in case testEquality ty ty' of
+                  Just Refl -> Right replace
+                  Nothing ->
+                    Left $
+                      TypeMismatch
+                        { typeMismatchName = nm
+                        , expectedType = C.Some ty
+                        , foundType = C.Some ty'
+                        }
+      Nothing -> Right s
+
+isInSubProg ::
+  -- | the program counter from the base of the target ELF
```

Comment:
I'm not sure how to understand this - Can you say more?

Reply:
Looks like the Haddock for fromDwarfInfo expands on this - perhaps refer to that?